lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,923 @@
+            req_by = "\n".join(
+                ["\t%s - '%s:%s'" % (to_text(p) if p else 'base', str(self), r) for p, r in self.required_by]
+            )
+            raise AnsibleError(
+                "%s from source '%s'. Available versions before last requirement added: %s\nRequirements from:\n%s"
+                % (msg, collection_source, ", ".join(self.versions), req_by)
+            )
+
+        self.versions = new_versions
+
+    def install(self, path, temp_path):
+        if self.skip:
+            display.display("Skipping '%s' as it is already installed" % str(self))
+            return
+
+        # Install if it is not
+        collection_path = os.path.join(path, self.namespace, self.name)
+        display.display("Installing '%s:%s' to '%s'" % (str(self), self.latest_version,
+                                                        collection_path))
+
+        if self.path is None:
+            download_url = self._galaxy_info['download_url']
+            artifact_hash = self._galaxy_info['artifact']['sha256']
+            self.path = _download_file(download_url, temp_path, artifact_hash, self._validate_certs)
+
+        if os.path.exists(collection_path):
+            shutil.rmtree(collection_path)
+        os.makedirs(collection_path)
+
+        with tarfile.open(self.path, mode='r') as collection_tar:
+            files_member_obj = collection_tar.getmember('FILES.json')
+            with _tarfile_extract(collection_tar, files_member_obj) as files_obj:
+                files = json.loads(to_text(files_obj.read()))
+
+            _extract_tar_file(collection_tar, 'MANIFEST.json', collection_path, temp_path)
+            _extract_tar_file(collection_tar, 'FILES.json', collection_path, temp_path)
+
+            for file_info in files['files']:
+                file_name = file_info['name']
+                if file_name == '.':
+                    continue
+
+                if file_info['ftype'] == 'file':
+                    _extract_tar_file(collection_tar, file_name, collection_path, temp_path,
+                                      expected_hash=file_info['chksum_sha256'])
+                else:
+                    os.makedirs(os.path.join(collection_path, file_name))
+
+    def set_latest_version(self):
+        self.versions = {self.latest_version}
+        self._get_metadata()
+
+    def _get_metadata(self):
+        if self._metadata:
+            return
+
+        collection_url = _urljoin(*[self.source, 'api', 'v2', 'collections', self.namespace, self.name, 'versions',
+                                    self.latest_version])
+        details = json.load(open_url(collection_url, validate_certs=self._validate_certs))
+        self._galaxy_info = details
+        self._metadata = details['metadata']
+
+    def _meets_requirements(self, version, requirements):
+        """
+        Supports version identifiers can be '==', '!=', '>', '>=', '<', '<=', '*'. Each requirement is delimited by ','
```

Comment:
Does this support `pkg_name >= 3, < 4, != 3.5`?

Reply:
In a very limited sense. Seems like `LooseVersion('1') < `LooseVersion('1.0') == True` and so on. So things like `>2` will work with packages `2.0`, `2.0.0`. This is pretty much governed by LooseVersion and if we want to expand on it it's probably best done in a separate PR. This one is already quite large.